Deprecation logger rejects new entries after upgrade to 9.0 unless deprecation log data stream is rolled over

### Elasticsearch Version

9.0

### Problem Description

In #117933, we changed the name of the dataset for the deprecation log data stream from `deprecation.elasticsearch` to `elasticsearch.deprecation`. If you had data in that data stream in 8.x or earlier, then when you upgrade to 9.0, if the deprecation logger is called the deprecations will be rejected. You will see entries like this in your (non-deprecation) logs:
```
Caused by: java.lang.IllegalArgumentException: [constant_keyword] field [data_stream.dataset] only accepts values that are equal to the value defined in the mappings [deprecation.elasticsearch], but got [elasticsearch.deprecation] 
```
```
[instance-0000000005] Bulk write of deprecation logs encountered some failures: [[4VK3j5UBamJ0EqvgUJmJ org.elasticsearch.index.mapper.DocumentParsingException: [1:87] failed to parse field [data_stream.dataset] of type [constant_keyword] in document with id '4VK3j5UBamJ0EqvgUJmJ'. Preview of field's value: 'elasticsearch.deprecation']]
```
The problem continues until you roll over the deprecation log data stream (`.logs-deprecation.default`). We discussed a couple of options to fix this:

1. A one-off service that gets started with a node from the deprecation xpack module. It would wait in a new feature in the cluster, and once that feature is available (so once all nodes are 9.0+), force a rollover of the `.logs-deprecation.default` datastream. Then the service would shut itself down.
2. Change the mapping of the [`dataset` field](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/template-resources/src/main/resources/deprecation/deprecation-indexing-mappings.json#L27) to be a `keyword` rather than a `constant_keyword`, since that field is no longer a constant in this data stream (it is a constant in any backing index, but across the data stream it can now have two different values).

### Steps to Reproduce

Start a cluster in 8.x. Do something that causes a deprecation message to be logged. Upgrade to 9.0. Do something that causes a deprecation message to be logged.

--- a/docs/changelog/125606.yaml
+++ b/docs/changelog/125606.yaml
@@ -1,0 +1,17 @@
+pr: 125606
+summary: Rename deprecation index template
+area: Infra/Logging
+type: breaking
+issues:
+ - 125445
+breaking:
+  title: Rename deprecation index template
+  area: Logging
+  details: The deprecation datastream contains log entries for deprecations that
+    occured while Elasticsearch is running. The deprecation log entries had a
+    mismatch in their dataset name. In order to avoid changing the dataset
+    name in the existing datastream, a new datastream now exists.
+  impact: If querying for deprecations previously using the
+    `.logs-deprecation.elasticsearch-default` datastream, you should now use the
+    `.logs-elasticsearch.deprecation-default` datastream.
+  notable: false

--- a/x-pack/plugin/core/template-resources/src/main/resources/deprecation/deprecation-indexing-template.json
+++ b/x-pack/plugin/core/template-resources/src/main/resources/deprecation/deprecation-indexing-template.json
@@ -1,5 +1,5 @@
 {
-  "index_patterns": [".logs-deprecation.*"],
+  "index_patterns": ["${xpack.deprecation.indexing.template.pattern}"],
   "priority": 1000,
   "data_stream": {
     "hidden": true
@@ -13,5 +13,6 @@
     "description": "default template for Stack deprecation logs index template installed by x-pack",
     "managed": true
   },
+  "deprecated": ${xpack.deprecation.indexing.template.deprecated},
   "version": ${xpack.deprecation.indexing.template.version}
 }

--- a/x-pack/plugin/deprecation/qa/common/src/main/java/org/elasticsearch/xpack/deprecation/DeprecationTestUtils.java
+++ b/x-pack/plugin/deprecation/qa/common/src/main/java/org/elasticsearch/xpack/deprecation/DeprecationTestUtils.java
@@ -23,7 +23,7 @@ public class DeprecationTestUtils {
     /**
      * Same as <code>DeprecationIndexingAppender#DEPRECATION_MESSAGES_DATA_STREAM</code>, but that class isn't visible from here.
      */
-    public static final String DATA_STREAM_NAME = ".logs-deprecation.elasticsearch-default";
+    public static final String DATA_STREAM_NAME = ".logs-elasticsearch.deprecation-default";
 
     static List<Map<String, Object>> getIndexedDeprecations(RestClient client) throws IOException {
         return getIndexedDeprecations(client, null);

--- a/x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingAppender.java
+++ b/x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingAppender.java
@@ -32,7 +32,7 @@ import java.util.function.Consumer;
 @Plugin(name = "DeprecationIndexingAppender", category = Core.CATEGORY_NAME, elementType = Appender.ELEMENT_TYPE)
 public class DeprecationIndexingAppender extends AbstractAppender {
     private static final Logger logger = LogManager.getLogger(DeprecationIndexingAppender.class);
-    public static final String DEPRECATION_MESSAGES_DATA_STREAM = ".logs-deprecation.elasticsearch-default";
+    public static final String DEPRECATION_MESSAGES_DATA_STREAM = ".logs-elasticsearch.deprecation-default";
 
     private final Consumer<IndexRequest> requestConsumer;
 

--- a/x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingComponent.java
+++ b/x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingComponent.java
@@ -137,7 +137,7 @@ public class DeprecationIndexingComponent extends AbstractLifecycleComponent imp
         }
         final IndexLifecycleMetadata indexLifecycleMetadata = event.state().metadata().custom(IndexLifecycleMetadata.TYPE);
 
-        if (event.state().getMetadata().templatesV2().containsKey(".deprecation-indexing-template")
+        if (event.state().getMetadata().templatesV2().containsKey(".deprecation-indexing-template-9")
             && indexLifecycleMetadata != null
             && indexLifecycleMetadata.getPolicies().containsKey(".deprecation-indexing-ilm-policy")) {
             flushEnabled.set(true);

--- a/x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingTemplateRegistry.java
+++ b/x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingTemplateRegistry.java
@@ -33,13 +33,15 @@ import static org.elasticsearch.xpack.core.ClientHelper.DEPRECATION_ORIGIN;
 public class DeprecationIndexingTemplateRegistry extends IndexTemplateRegistry {
     // history (please add a comment why you increased the version here)
     // version 1: initial
-    public static final int INDEX_TEMPLATE_VERSION = 1;
+    // version 2: deprecated old name and renamed index pattern
+    public static final int INDEX_TEMPLATE_VERSION = 2;
 
     public static final String DEPRECATION_INDEXING_TEMPLATE_VERSION_VARIABLE = "xpack.deprecation.indexing.template.version";
 
     public static final String DEPRECATION_INDEXING_MAPPINGS_NAME = ".deprecation-indexing-mappings";
     public static final String DEPRECATION_INDEXING_SETTINGS_NAME = ".deprecation-indexing-settings";
-    public static final String DEPRECATION_INDEXING_TEMPLATE_NAME = ".deprecation-indexing-template";
+    public static final String DEPRECATION_INDEXING_TEMPLATE_NAME_OLD = ".deprecation-indexing-template";
+    public static final String DEPRECATION_INDEXING_TEMPLATE_NAME = ".deprecation-indexing-template-9";
     public static final String DEPRECATION_INDEXING_POLICY_NAME = ".deprecation-indexing-ilm-policy";
 
     public DeprecationIndexingTemplateRegistry(
@@ -89,7 +91,25 @@ public class DeprecationIndexingTemplateRegistry extends IndexTemplateRegistry {
             DEPRECATION_INDEXING_TEMPLATE_NAME,
             "/deprecation/deprecation-indexing-template.json",
             INDEX_TEMPLATE_VERSION,
-            DEPRECATION_INDEXING_TEMPLATE_VERSION_VARIABLE
+            DEPRECATION_INDEXING_TEMPLATE_VERSION_VARIABLE,
+            Map.of(
+                "xpack.deprecation.indexing.template.deprecated",
+                "false",
+                "xpack.deprecation.indexing.template.pattern",
+                ".logs-elasticsearch.deprecation-*"
+            )
+        ),
+        new IndexTemplateConfig(
+            DEPRECATION_INDEXING_TEMPLATE_NAME_OLD,
+            "/deprecation/deprecation-indexing-template.json",
+            INDEX_TEMPLATE_VERSION,
+            DEPRECATION_INDEXING_TEMPLATE_VERSION_VARIABLE,
+            Map.of(
+                "xpack.deprecation.indexing.template.deprecated",
+                "true",
+                "xpack.deprecation.indexing.template.pattern",
+                ".logs-deprecation.*"
+            )
         )
     );
 

